### PR TITLE
Fixing 2176

### PR DESCRIPTION
**Pull request by biddisco**
These patches solve the problem of launching the test correctly, but do not address the memory access violations that occur. These could be merged in, independently of the greater fix.

I do not know how to progress with the sanitizer errors and double deletions mentioned in #2176

### CODE REVIEW

**Comment by hkaiser** [comment]
I'd have preferred to have a proper fix for the ipv6 address comparison instead of just removing them. Could you please try to explain what's wrong and perhaps what would have to be done in order to properly fix this?

**Comment by biddisco** [comment]
The ipv6 comparison is actually just comparing strings. the correct fix would be to convert them back into ipv6 addresses and compare the binary versions. I implemented this, but the fix was specific for ipv6 and would fail for ipv4, so we need to check what type of addresses they are first and convert them accordingly. Since the underlying address objects that provided the strings must be available somewhere, I decided not to spend time on thins ans just comment the checks out. As you mentioned, they have never triggered until now and are not serving any useful purpose.

**Comment by biddisco** [comment]
@hkaiser I have added a cleanup_ip_address function that takes a string containing an ipv4 or ipv6 address and attempts to convert it into a "standardized" address string. It's a bit of a waste of CPU, but it is only called in Debug mode. I hope that addresses your concerns.

File: src/util/asio_util.cpp [hunk 1]
```diff
@@ -172,6 +172,27 @@ namespace hpx { namespace util
         return true;
     }
 
+    ///////////////////////////////////////////////////////////////////////
+    // Take an ip v4 or v6 address and "standardize" it for comparison checks
+    HPX_API_EXPORT std::string cleanup_ip_address(const std::string &addr)
```
Comment (hkaiser) [inline src/util/asio_util.cpp]: Export attribute are needed on the first declaration only. gcc will issue warnings for repeating it on the definition.

File: src/util/asio_util.cpp [hunk 2]
```diff
@@ -172,6 +172,27 @@ namespace hpx { namespace util
...
+        unsigned char buf[sizeof(struct in6_addr)];
+        int i=0, domain[2] = {AF_INET, AF_INET6};
+        char str[INET6_ADDRSTRLEN];
+
+        for (i=0; i<2; ++i) {
+            int s = inet_pton(domain[i], &addr[0], buf);
+            if (s>0) break;
+        }
+        if (i==2) {
+            throw std::runtime_error("Invalid IP address string");
```
Comment (hkaiser) [inline src/util/asio_util.cpp]: Please throw exceptions using `HPX_THROW_EXCEPTION()`

**Comment by hkaiser** [comment]
@biddisco Thanks, this looks fine. I think we shouldn't worry about wasting cycles in Debug code.

**Comment by biddisco** [comment]
@hkaiser I realized that the inet_xxx functions would not compile under windows, so I have added  a couple of (untested) snippets I found on stack overflow. I'm on the train and my connection is poor - coulf you trigger a windows build of this and I'll check results later.

**Comment by hkaiser** [comment]
@biddisco see http://hermione.cct.lsu.edu/builders/hpx_msvc_12_boost_1_57_windows_x86_64_debug/builds/786 and http://hermione.cct.lsu.edu/builders/hpx_msvc_14_boost_1_59_dev_windows_x86_64_debug/builds/490

File: src/util/asio_util.cpp [hunk 3]
```diff
@@ -172,6 +172,57 @@ namespace hpx { namespace util
...
+                break;
+            default:
+                return NULL;
+        }
+        /* cannot direclty use &size because of strict aliasing rules */
+        return
+        (WSAAddressToString((struct sockaddr *)&ss, sizeof(ss), NULL, dst, &s) == 0)?
+                dst : NULL;
+    }
+#endif
```
Comment (hkaiser) [inline src/util/asio_util.cpp]: @biddisco These functions are available from ASIO: see `boost::asio::detail::socket_ops::inet_pton` and `boost::asio::detail::socket_ops::inet_ntop`.

**Comment by hkaiser** [comment]
LGTM, thanks!

**Comment by biddisco** [comment]
Don't merge this. I think there is a problem with the ip address checks and i have not got around to looking at it again. Once I've double checked it, I'll report back.

**Comment by biddisco** [comment]
I spent some time experimenting and have found that on my mac when I have the VPN to the office open, I get ipv6 addresses that are incorrectly compared by the current code - and this causes the test to fail. 
The patches included do a ipv6 cleanup, but there's a catch. When I use the boost::asio::detail::socket_ops::inet_pton and pton functions - they are too clever and convert the ip address to and from the binary descrition, preserving the %en0 suffix (link-local zone index - I don't fully understand it).
if I use the standard inet_pton and inet_ntop I get this
```
cleanup 1 fe80:5::bae8:56ff:fe40:c3d0   = fe80:5::bae8:56ff:fe40:c3d0
cleanup 2 fe80::bae8:56ff:fe40:c3d0%en0 = fe80:5::bae8:56ff:fe40:c3d0
```
but when I use the boost asio functions, I get
```
cleanup 1 fe80:5::bae8:56ff:fe40:c3d0   = fe80:5::bae8:56ff:fe40:c3d0
cleanup 2 fe80::bae8:56ff:fe40:c3d0%en0 = fe80::bae8:56ff:fe40:c3d0%en0
```
So I am leaving the linux inet_pton functions in, but using the boost asio functions for windows and have added a comment in the source code pointing to this #issue in the event that anyone ever find a problem on windows with ipv6 addresses and looks at it.

In summary : without these patches, the test fails for me on OSX (when the VPN is used and we have an ipv6 address) . With them it passes.

**Comment by hkaiser** [comment]
Thanks John!

**Comment by biddisco** [comment]
It says the checks failed, but actually they passed and the build timed out just before finishing.
